connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/TestPlugins.java

```diff
@@ -0,0 +1,241 @@
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.connect.runtime.isolation;
+
+import java.io.BufferedInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.StringWriter;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.jar.Attributes;
+import java.util.jar.JarEntry;
+import java.util.jar.JarOutputStream;
+import java.util.jar.Manifest;
+import java.util.stream.Collectors;
+import javax.tools.JavaCompiler;
+import javax.tools.StandardJavaFileManager;
+import javax.tools.ToolProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Utility class for constructing test plugins for Connect.
+ *
+ * <p>Plugins are built from their source under resources/test-plugins/ and placed into temporary
+ * jar files that are deleted when the process exits.
+ *
+ * <p>To add a plugin, create the source files in the resource tree, and edit this class to build
+ * that plugin during initialization. For example, the plugin class {@literal package.Class} should
+ * be placed in {@literal resources/test-plugins/something/package/Class.java} and loaded using
+ * {@code createPluginJar("something")}. The class name, contents, and plugin directory can take
+ * any value you need for testing.
+ *
+ * <p>To use this class in your tests, make sure to first call
+ * {@link TestPlugins#assertInitialized()} to verify that the plugins initialized correctly.
+ * Otherwise, exceptions during the plugin build are not propagated, and may invalidate your test.
+ * You can access the list of plugin jars for assembling a {@literal plugin.path}, and reference
+ * the names of the different plugins directly via the exposed constants.
+ */
+public class TestPlugins {
+
+    /**
+     * Class name of a plugin which will always throw an exception during loading
+     */
+    public static final String ALWAYS_THROW_EXCEPTION = "test.plugins.AlwaysThrowException";
```

Comment:
curious. Is there an issue if we load those early? (and therefore get their name without risking typos?)

Reply:
We can't reference this as `AlwaysThrowExceptions.class.getName()`, because these plugins live in the resources directory, and aren't really recognized as compilable classes. They can't be included as a compilable class in the system class loader, because then we wouldn't be able to adequately test the isolation.
We could determine the class name by scanning for `Converter` implementations inside the plugins directory, but then this adds more complication. It would essentially duplicate logic inside Kafka itself, which may be under test in the first place.

Reply:
I'd like to insist a bit here, so that we are quite deliberate in terms of what is loaded and when. Isolation aims to isolate class loading even from the system classloader. When you use the delegating classloader in your tests, isn't such isolation happening?

Reply:
If these plugin classes were included as normal classes, they would be part of the system classloader. The tests would then also need to verify that we were not falling back to the system classloader, where under the current system, these classes are not present, and any fallback would thrown an error.

We would need to also produce a jar for creating the `PluginClassLoader`, and I'm not sure how to extract additional jars from the main source directory for use in the `PluginClassLoader`s.

Reply:
If you have an instance of a class object you can get the classloader that loaded the class. This way you could verify whether it was found from the plugin.path or not. 

Haven't checkout the code to see if there are any other hoops you'd need to jump to make reasonable assertions on your tests. But this could be a way. Again, it's not only about using Class objects instead of String objects but also to make the tests more robust by making sure that a class is verified as found by a plugin classloader and not the system one. Just another thought, but we've already made good progress with these tests

Reply:
In order to verify that the PluginClassLoader is active, these checks are already being done. If these strings are wrong, a ClassNotFound exception will be thrown during the tests and it won't go unnoticed. I don't think there are any benefits here for getting these classnames to typecheck.

Reply:
Thanks. Not a blocker.